README.md
# Brightway2-analyzer

[![PyPi Version](https://img.shields.io/pypi/v/bw2analyzer.svg)](https://pypi.org/project/bw2analyzer/)
[![Conda Version](https://img.shields.io/conda/vn/conda-forge/bw2analyzer.svg)](https://anaconda.org/conda-forge/bw2analyzer)

This is a set of tools to analyze results of life-cycle assessment (LCA) calculations. It is part of the [Brightway2 LCA framework](https://brightway.dev).


CHANGES.md
# `bw2analyzer` Changelog

## 0.11.7 (2023-04-25)

Remove `pyprind` dependency

## 0.11.6 (2023-04-25)

Remove incorrect dependencies

## 0.11.5 (2023-04-24)

* [#23](https://github.com/brightway-lca/brightway2-analyzer/pull/23): Remove the length limit on the printed strings
* [#22](https://github.com/brightway-lca/brightway2-analyzer/pull/22): Fix contribution analysis error
* [#21](https://github.com/brightway-lca/brightway2-analyzer/pull/21): Badges

## 0.11.4 (2022-07-04)

* Add `use_matrix_values` to `recursive_calculation_to_object` and `print_recursive_calculation` for Monte Carlo
* Add passing custom LCA classes to `recursive_calculation_to_object` and `print_recursive_calculation` for Monte Carlo and regionalized LCA

## 0.11.3 (2022-06-22)

* Add `recursive_calculation_to_object` function which returns lists or pandas DataFrames.

## 0.11.2 (2022-04-01)

* [PR #19: Fixed negative contributions](https://github.com/brightway-lca/brightway2-analyzer/pull/19). Thanks @nabilahmed739.
* [PR #12: Add helper function to label life cycle inventory](https://github.com/brightway-lca/brightway2-analyzer/pull/12). Thanks @BenPortner.
* [PR #11: Add biosphere activity information to `multi_recurse_tagged_database`](https://github.com/brightway-lca/brightway2-analyzer/pull/11). Thanks Thanks @BenPortner.
* [PR #10: Include biosphere activity information in tagged graph traversal ](https://github.com/brightway-lca/brightway2-analyzer/pull/10). Thanks @BenPortner.

## 0.11.1 (2021-10-24)

* Fix bug in [non-unitary production amounts for recursive calculations](https://github.com/brightway-lca/from-the-ground-up/issues/1)

## 0.11 (2021-10-21)

* Brightway 2.5 compatibility
* Removed `DatabaseExplorer`
* Removed `ParameterFinder`
* Deprecated `SerializedLCAReport`

## 0.10 (2021-05-06)

* Add `compare_activities_by_grouped_leaves` function.
* Add `find_differences_in_inputs` function.
* Add `compare_activities_by_lcia_score` function.
* Add `print_recursive_calculation` and `print_recursive_supply_chain` utility functions.
* Add dependencies on `pandas` and [tabulate](https://pypi.org/project/tabulate/).

### 0.9.5 (2019-11-06)

* Merge [PR #2](https://bitbucket.org/cmutel/brightway2-analyzer/pull-requests/2/multiple-methods-secondary-tags-and/commits), allowing for more powerful tagged graph traversal
* Fix error in nonunitary production for ``traverse_tagged_databases``.

### 0.9.4 (2017-04-17)

* Fix license text

### 0.9.3 (2017-04-11)

* Cleaned up setup and license

### 0.9.1 (2016-07-14)

* Compatibility with bw2data 2.3

## 0.9 (2016-04-09)

* Py3 compatibility
* Updates for compatibility with downstream changes

## 0.8 (2015-03-07)

* Compatibility with bw2calc 1.0 (split products and activities)
* Add ordered matrix graph

## 0.7 (2014-12-09)

Thanks Bernhard Steubing for good ideas and bugfixes.

* ENHANCEMENT: Add amount to annotated top processes and emissions.
* BUGFIX: Propagate **kwargs** in contribution analysis.

### 0.6.1 (2014-08-30)

* ENHANCMENT: Add ouroboros health check (processes which consume their own outcome)

## 0.6 (2014-08-29)

* FEATURE: Database health check! Does a number of tests on a Database to see if potential problems can be identified.

### 0.5.2 (2014-07-30)

Updated dependencies.

### 0.5.1 (2014-04-16)

* CHANGE: ContributionAnalysis.annotated_top_* will automatically reverse LCA object dictionaries if needed.

## 0.5 (2014-01-29)

* CHANGE: Compatibility fixes with bw2data 0.12 and bw2calc 0.11


bw2analyzer/tagged.py
from collections import defaultdict
from warnings import warn
import pandas as pd
from bw2calc import LCA
from bw2data import Method, get_activity, Database
from typing import List, Union


def traverse_tagged_databases(
    functional_unit, method, label="tag", default_tag="other", secondary_tags=[], fg_databases=None
):

    """Traverse a functional unit throughout its foreground database(s) or the 
    listed databses in fg_databses, and group impacts by tag label.

    Contribution analysis work by linking impacts to individual activities.
    However, you also might want to group impacts in other ways. For example,
    give individual biosphere exchanges their own grouping, or aggregate two
    activities together.

    Consider this example system, where the letters are the tag labels, and the
    numbers are exchange amounts. The functional unit is one unit of the tree
    root.

    .. image:: images/tagged-traversal.png
       :alt: Example tagged supply chain

    In this supply chain, tags are applied to activities and biosphere exchanges.
    If a biosphere exchange is not tagged, it inherits the tag of its producing
    activity. Similarly, links to other databases are assessed with the usual
    LCA machinery, and the total LCA score is tagged according to its consuming
    activity. If an activity does not have a tag, a default tag is applied.

    We can change our visualization to show the use of the default tags:

    .. image:: images/tagged-traversal-2.png
       :alt: Example tagged supply chain

    And then we can manually calculate the tagged impacts. Normally we would
    need to know the actual biosphere flows and their respective
    characterization factors (CF), but in this example we assume that each
    CF is one. Our result, group by tags, would therefore be:

        * **A**: :math:`6 + 27 = 33`
        * **B**: :math:`30 + 44 = 74`
        * **C**: :math:`5 + 16 + 48 = 69`
        * **D**: :math:`14`

    This function will only traverse the foreground database, i.e. the
    database of the functional unit activity. A functional unit can have
    multiple starting nodes; in this case, all foreground databases are
    traversed.

    Input arguments:

        * ``functional_unit``: A functional unit dictionary, e.g. ``{("foo", "bar"): 42}``.
        * ``method``: A method name, e.g. ``("foo", "bar")``
        * ``label``: The label of the tag classifier. Default is ``"tag"``
        * ``default_tag``: The tag classifier to use if none was given. Default is ``"other"``
        * ``secondary_tags``: List of tuples in the format (secondary_label, secondary_default_tag). Default is empty list.
        * ``fg_databases``: a list of foreground databases to be traversed, e.g. ['foreground', 'biomass', 'machinery']
                            It's not recommended to include all databases of a project in the list to be traversed, especially not ecoinvent itself

    Returns:

        Aggregated tags dictionary from ``aggregate_tagged_graph``, and tagged supply chain graph from ``recurse_tagged_database``.

    """

    lca = LCA(functional_unit, method)
    lca.lci()
    lca.lcia()

    method_dict = {o[0]: o[1] for o in Method(method).load()}

    graph = [
        recurse_tagged_database(
            key, amount, method_dict, lca, label, default_tag, secondary_tags, fg_databases
        )
        for key, amount in functional_unit.items()
    ]

    return aggregate_tagged_graph(graph), graph


def aggregate_tagged_graph(graph):
    """Aggregate a graph produced by ``recurse_tagged_database`` by the provided tags.

    Outputs a dictionary with keys of tags and numeric values.

    .. code-block:: python

        {'a tag': summed LCIA scores}

    """

    def recursor(obj, scores):
        scores[obj["tag"]] += obj["impact"]
        for flow in obj["biosphere"]:
            scores[flow["tag"]] += flow["impact"]
        for exc in obj["technosphere"]:
            scores = recursor(exc, scores)
        return scores

    scores = defaultdict(int)
    for obj in graph:
        scores = recursor(obj, scores)
    return scores


def flatten_tagged_graph(graph:dict, as_dataframe:bool=True) -> Union[List[dict],pd.DataFrame]:
    # convert nested graph into flat edge list: [{source, target, other properties}]

    def flatten_recursive(graph, target=None, agg=0):
        # add activity metadata
        d = graph["activity"]._data
        # add other graph data
        for k,v in graph.items():
            if k in ["technosphere", "biosphere"]:
                continue
            else:
                d[k] = v
        d["source"] = len(flat_list)
        d["target"] = target
        flat_list.append(d)
        # get upstream contributions
        for ex in graph.get("biosphere", []) + graph.get("technosphere", []):
            agg += flatten_recursive(ex, target=d["source"])
        # calculate absolute impact (direct + upstream)
        d["value"] = agg + d["impact"]
        return d["value"]

    flat_list = []
    flatten_recursive(graph)

    if as_dataframe is True:
        return pd.DataFrame(flat_list).convert_dtypes({"target": pd.Int64Dtype})
    else:
        return flat_list


def recurse_tagged_database(
    activity, amount, method_dict, lca, label, default_tag, secondary_tags=[], fg_databases=None, warned=False
):

    """Traverse a foreground database and assess activities and biosphere flows by tags.


    Input arguments:

        * ``activity``: Activity tuple or object
        * ``amount``: float
        * ``method_dict``: Dictionary of biosphere flow tuples to CFs, e.g. ``{("biosphere", "foo"): 3}``
        * ``lca``: An ``LCA`` object that is already initialized, i.e. has already calculated LCI and LCIA with same method as in ``method_dict``
        * ``label``: string
        * ``default_tag``: string
        * ``secondary_tags``: List of tuples in the format (secondary_label, secondary_default_tag). Default is empty list.
        
        * ``fg_databases``: a list of foreground databases to be traversed, e.g. ['foreground', 'biomass', 'machinery']
                            It's not recommended to include all databases of a project in the list to be traversed, especially not ecoinvent itself

    Returns:

    .. code-block:: python

        {
            'activity': activity object,
            'amount': float,
            'tag': string,
            'secondary_tags': [list of strings],
            'impact': float (impact of inputs from outside foreground database),
            'biosphere': [{
                'amount': float,
                'impact': float,
                'tag': string,
                'secondary_tags': [list of strings]
            }],
            'technosphere': [this data structure]
        }

    """
    if isinstance(activity, tuple):
        activity = get_activity(activity)

    MESSAGE = """Given databases include many activities, and traversal may be slow.
Consider using `GraphTraversalLCA` from `bw2calc` instead."""

    if fg_databases is None:  # then set the list equal to the database of the functional unit
        fg_databases = [activity['database']]  # list, single item
    elif not warned and sum(len(Database(name)) for name in fg_databases) > 2500:
        warn(MESSAGE)
        warned = True

    inputs = list(activity.technosphere())
    production = list(activity.production())

    if not production:
        scale = 1
    elif len(production) > 1:
        warn("Hit multiple production exchanges; aborting in this branch")
        return
    else:
        scale = production[0]["amount"]
        for other in activity.technosphere():
            if other.input == production[0].input:
                scale -= other["amount"]

    inside = [exc for exc in inputs if exc.input["database"] in fg_databases]

    outside = {
        exc.input.id: exc["amount"] / scale * amount
        for exc in inputs
        if exc["input"][0] not in fg_databases
    }

    if outside:
        lca.redo_lcia(outside)
        outside_score = lca.score
    else:
        outside_score = 0

    return {
        "activity": activity,
        "amount": amount,
        "tag": activity.get(label) or default_tag,
        "secondary_tags": [activity.get(t[0]) or t[1] for t in secondary_tags],
        "impact": outside_score,
        "biosphere": [
            {
                "activity": exc.input,
                "amount": exc["amount"] / scale * amount,
                "impact": exc["amount"]
                / scale
                * amount
                * method_dict.get(exc["input"], 0),
                "tag": exc.get(label) or activity.get(label) or default_tag,
                "secondary_tags": [
                    exc.get(t[0]) or activity.get(t[0]) or t[1] for t in secondary_tags
                ],
            }
            for exc in activity.biosphere()
        ],
        "technosphere": [
            recurse_tagged_database(
                activity=exc.input,
                amount=exc["amount"] / scale * amount,
                method_dict=method_dict,
                lca=lca,
                label=label,
                default_tag=default_tag,
                secondary_tags=secondary_tags,
                fg_databases=fg_databases,
                warned=warned,
            )
            for exc in inside
        ],
    }


## tagged graph functions using multiple methods


def multi_traverse_tagged_databases(
    functional_unit, methods, label="tag", default_tag="other", secondary_tags=[]
):

    """Traverse a functional unit throughout its foreground database(s), and
    group impacts (for multiple methods) by tag label.

    Input arguments:
        * ``functional_unit``: A functional unit dictionary, e.g. ``{("foo", "bar"): 42}``.
        * ``methods``: A list of method names, e.g. ``[("foo", "bar"), ("baz", "qux"), ...]``
        * ``label``: The label of the tag classifier. Default is ``"tag"``
        * ``default_tag``: The tag classifier to use if none was given. Default is ``"other"``
        * ``secondary_tags``: List of tuples in the format (secondary_label, secondary_default_tag). Default is empty list.

    Returns:

        Aggregated tags dictionary from ``aggregate_tagged_graph``, and tagged supply chain graph from ``recurse_tagged_database``.

    """

    lca = LCA(functional_unit, methods[0])
    lca.lci()  # factorize=True)
    lca.lcia()

    method_dicts = [{o[0]: o[1] for o in Method(method).load()} for method in methods]

    graph = [
        multi_recurse_tagged_database(
            key, amount, methods, method_dicts, lca, label, default_tag, secondary_tags
        )
        for key, amount in functional_unit.items()
    ]

    return multi_aggregate_tagged_graph(graph), graph


def multi_aggregate_tagged_graph(graph):

    """Aggregate a graph produced by ``multi_recurse_tagged_database`` by the provided tags.

    Outputs a dictionary with keys of tags and numeric values.

    Note: this only aggregates on the primary tag, secondary tags are not aggregated

    .. code-block:: python

        {'a tag': [list of summed LCIA scores with one sum per method]}

    """

    def recursor(obj, scores):
        if not scores.get(obj["tag"]):
            scores[obj["tag"]] = [x for x in obj["impact"]]
        else:
            scores[obj["tag"]] = [
                sum(x) for x in zip(scores[obj["tag"]], obj["impact"])
            ]

        for flow in obj["biosphere"]:
            if not scores.get(flow["tag"]):
                scores[flow["tag"]] = [x for x in flow["impact"]]
            else:
                scores[flow["tag"]] = [
                    sum(x) for x in zip(scores[flow["tag"]], flow["impact"])
                ]

        for exc in obj["technosphere"]:
            scores = recursor(exc, scores)
        return scores

    scores = defaultdict(int)
    for obj in graph:
        scores = recursor(obj, scores)
    return scores


def multi_recurse_tagged_database(
    activity, amount, methods, method_dicts, lca, label, default_tag, secondary_tags=[]
):

    """Traverse a foreground database and assess activities and biosphere flows by tags using multiple methods.

    Input arguments:

        * ``activity``: Activity tuple or object
        * ``amount``: float
        * ``methods``: list of LCA methods (tuples)
        * ``method_dicts``: list of dictionaries of biosphere flow tuples to CFs, e.g. ``{("biosphere", "foo"): 3}`` corresponding to methods in ``methods``
        * ``lca``: An ``LCA`` object that is already initialized, i.e. has already calculated LCI
        * ``label``: string
        * ``default_tag``: string
        * ``secondary_tags``: list of tuples in the format (secondary_label, secondary_default_tag). Default is empty list.

    Returns:

    .. code-block:: python

        {
            'activity': activity object,
            'amount': float,
            'tag': string,
            'secondary_tags': [list of strings],
            'impact': [list of floats (impact of inputs from outside foreground database) with one element per method],
            'biosphere': [{
                'amount': float,
                'impact': [list of floats with one element per method],
                'tag': string,
                'secondary_tags': [list of strings]
            }],
            'technosphere': [this data structure]
        }

    """

    if isinstance(activity, tuple):
        activity = get_activity(activity)

    inputs = list(activity.technosphere())
    inside = [exc for exc in inputs if exc.input["database"] == activity["database"]]
    outside = {
        exc.input.id: exc["amount"] * amount
        for exc in inputs
        if exc["input"][0] != activity["database"]
    }

    if outside:
        outside_scores = []
        for n, m in enumerate(methods):
            lca.switch_method(m)
            lca.redo_lcia(outside)
            outside_scores.append(lca.score)
    else:
        outside_scores = [0] * len(methods)

    return {
        "activity": activity,
        "amount": amount,
        "tag": activity.get(label) or default_tag,
        "secondary_tags": [activity.get(t[0]) or t[1] for t in secondary_tags],
        "impact": outside_scores,
        "biosphere": [
            {
                "activity": exc.input,
                "amount": exc["amount"] * amount,
                "impact": [
                    exc["amount"] * amount * method_dict.get(exc["input"], 0)
                    for method_dict in method_dicts
                ],
                "tag": exc.get(label) or activity.get(label) or default_tag,
                "secondary_tags": [
                    exc.get(t[0]) or activity.get(t[0]) or t[1] for t in secondary_tags
                ],
            }
            for exc in activity.biosphere()
        ],
        "technosphere": [
            multi_recurse_tagged_database(
                exc.input,
                exc["amount"] * amount,
                methods,
                method_dicts,
                lca,
                label,
                default_tag,
                secondary_tags,
            )
            for exc in inside
        ],
    }


def get_cum_impact(graph, max_levels=100):

    """Add cumulative impact ``cum_impact`` to each ``technosphere`` level of a tagged graph.

    This function recurses until all levels in the graph have been checked, or the ``max_levels`` cutoff is reached

    Input arguments:
        * ``graph``: A tagged supply chain graph from ``recurse_tagged_database``.
        * ``max_levels``: maximum number of graph levels to check before giving up. Default is 100.

    Returns:
         Tagged supply chain graph with additional cumulative impact ``cum_impact`` key at each ``technosphere`` level.
    """

    def cum_impact_recurse(d):

        to_return = {}
        cum_impact = 0

        for k, v in d.items():
            if k == "technosphere":
                if len(v) != 0:
                    for e in v:
                        cum_impact += e["impact"]
                        if "cum_impact" in e.keys():
                            cum_impact += e["cum_impact"]

                        if k in to_return.keys():
                            to_return[k].append(cum_impact_recurse(e))
                        else:
                            to_return[k] = [cum_impact_recurse(e)]
                else:
                    to_return[k] = []

            elif k == "biosphere":
                to_return[k] = v
                if len(v) != 0:
                    for b in v:
                        cum_impact += b["impact"]

            # elif k == 'activity':
            #    to_return[k] = str(v)

            else:
                to_return[k] = v

        to_return["cum_impact"] = cum_impact

        return to_return

    return_list = []

    for subgraph in graph:

        this_d = subgraph

        for i in range(max_levels):
            prev_d = this_d
            this_d = cum_impact_recurse(prev_d)
            if this_d == prev_d:
                break

        return_list.append(this_d)

    return return_list


def get_multi_cum_impact(graph, max_levels=100):

    """Add cumulative impact ``cum_impact`` to each ``technosphere`` level of a multi method tagged graph.

    This function recurses until all levels in the graph have been checked, or the ``max_levels`` cutoff is reached

    Input arguments:
        * ``graph``: A tagged supply chain graph from ``multi_recurse_tagged_database``.
        * ``max_levels``: maximum number of graph levels to check before giving up. Default is 100.

    Returns:
         Tagged supply chain graph with additional cumulative impact ``cum_impact`` key at each ``technosphere`` level.
    """

    def multi_cum_impact_recurse(d):

        to_return = {}
        cum_impact = [0] * len(d["impact"])

        for k, v in d.items():
            if k == "technosphere":
                if len(v) != 0:
                    for e in v:
                        cum_impact = [sum(x) for x in zip(cum_impact, e["impact"])]
                        if "cum_impact" in e.keys():
                            cum_impact = [
                                sum(x) for x in zip(cum_impact, e["cum_impact"])
                            ]

                        if k in to_return.keys():
                            to_return[k].append(multi_cum_impact_recurse(e))
                        else:
                            to_return[k] = [multi_cum_impact_recurse(e)]
                else:
                    to_return[k] = []

            elif k == "biosphere":
                to_return[k] = v
                if len(v) != 0:
                    for b in v:
                        cum_impact = [sum(x) for x in zip(cum_impact, b["impact"])]

            # elif k == 'activity':
            #    to_return[k] = str(v)

            else:
                to_return[k] = v
        to_return["cum_impact"] = cum_impact

        return to_return

    return_list = []

    for subgraph in graph:

        this_d = subgraph

        for i in range(max_levels):
            prev_d = this_d
            this_d = multi_cum_impact_recurse(prev_d)
            if this_d == prev_d:
                break

        return_list.append(this_d)

    return return_list
